rest_witchcraft/mixins.py

```diff
@@ -12,3 +19,108 @@ class DestroyModelMixin(mixins.DestroyModelMixin):
     def perform_destroy(self, instance):
         session = self.get_session()
         session.delete(instance)
+
+
+class QuerySerializerMixin(object):
+    """
+    Adds query serializer validation logic to viewset
+
+    Query will be validated as part of query viewset initialization
+    therefore query will be validated before any of the viewset actions
+    are executed.
+
+    In addition query serializer will be included in serializer context
+    for standard viewset serializers. That
+    """
+
+    query_serializer_class = None
+
+    @property
+    def query_serializer(self):
+        return getattr(self, "_query_serializer", None)
+
+    @query_serializer.setter
+    def query_serializer(self, value):
+        self._query_serializer = value
+
+    def get_query_serializer_class(self):
+        return (
+            self.query_serializer_class
+            or getattr(self.get_serializer_class()(), "get_query_serializer_class", lambda: None)()
```

Comment:
not a fan of initializing serializer here but its necessary as otherwise none of the nested fields are considered. hopefully people dont hit db, etc when initializing serializer....

initially was doing it on class itself however that only considered declared fields therefore any autogenerated fields which `ModelSerializer` can automatically add were completely ignored.